lib/social_share_button_helper.rb

```diff
@@ -0,0 +1,48 @@
+module SocialShareButtonHelper
+  require "uri"
+
+  SOCIAL_SHARE_CONFIG = {
+    :email => "social_icons/email.svg",
+    :facebook => "social_icons/facebook.svg",
+    :linkedin => "social_icons/linkedin.svg",
+    :mastodon => "social_icons/mastodon.svg",
+    :telegram => "social_icons/telegram.svg",
+    :x => "social_icons/x.svg"
+  }.freeze
+
+  def self.filter_allowed_sites(sites)
+    valid_sites = sites.empty? ? SOCIAL_SHARE_CONFIG.keys : sites.select { |site| valid_site?(site) }
+    invalid_sites = sites - valid_sites
+    [valid_sites, invalid_sites]
+  end
+
+  def self.icon_path(site)
+    SOCIAL_SHARE_CONFIG[site.to_sym] || ""
+  end
+
+  def self.valid_site?(site)
+    SOCIAL_SHARE_CONFIG.key?(site.to_sym)
+  end
+
+  def self.generate_share_url(site, params)
+    site = site.to_sym
+    case site
+    when :email
+      "mailto:?subject=#{URI.encode_www_form_component(params[:title])}&body=#{URI.encode_www_form_component(params[:url])}"
```

Comment:
`URI.encode_www_form_component` is a bit different from javascript's `encodeURIComponent`. It encodes spaces as `+`. But [mailto rfc](https://datatracker.ietf.org/doc/html/rfc6068#section-5) says spaces should be encoded as `%20`.